src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,278 @@
-    super.onActivityStarted(activity);
+  @Override
+  public void identify(IdentifyPayload identify) {
+    super.identify(identify);
 
+    if (!isNullOrEmpty(identify.userId())) {
+      firebaseAnalytics.setUserId(identify.userId());
+      Map<String, Object> traits = identify.traits();
+      for (Map.Entry<String, Object> entry : traits.entrySet()) {
+        String trait = entry.getKey();
+        trait = trait.trim().replaceAll(" ", "_");
+        if (trait.length() > 40) {
+          trait = trimKey(trait);
+        }
+        String formattedValue;
+        if (entry.getValue() instanceof Date) {
+          Date value = (Date) entry.getValue();
+          formattedValue = formatDate(value);
+        } else if (entry.getValue() instanceof String && ISO_DATE.matcher(String.valueOf(entry.getValue())).matches()) {
+          formattedValue = String.valueOf(entry.getValue()).substring(0, 10);
+        } else {
+          formattedValue = String.valueOf(entry.getValue());
+        }
+
+        firebaseAnalytics.setUserProperty(trait, formattedValue);
+        logger.verbose("firebaseAnalytics.setUserProperty(%s, %s);", trait, formattedValue);
+      }
+    }
   }
 
-  @Override public void onActivityStopped(Activity activity) {
-    super.onActivityStopped(activity);
+  @Override
+  public void screen(ScreenPayload screen) {
+    super.screen(screen);
 
+    logger.verbose(
+            "The Firebase SDK gathers screen calls automatically. Segment does map manual screen " +
+                    "methods to Firebase.");
   }
 
-  @Override public void identify(IdentifyPayload identify) {
-    super.identify(identify);
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
 
-    if (!isNullOrEmpty(identify.userId())) {
-      mFirebaseAnalytics.setUserId(identify.userId());
+    String event = track.event();
+    String eventName = mapEvent(event);
+    Properties properties = track.properties();
+
+    try {
+      Bundle formattedProperties = formatProperties(properties);
+      firebaseAnalytics.logEvent(eventName, formattedProperties);
+      logger.verbose("firebaseAnalytics.logEvent(%s, %s);", eventName, formattedProperties);
+    } catch (Exception e) {
+      logger.verbose(
+              "You must set `currency` in your event's property object to accurately "
+                      + "pass 'value' to Firebase. This event was not sent to Firebase.");
     }
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
+  private String mapEvent(String event) {
+    String eventName = event;
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
+    }
+    if (eventName.contains(" ")) {
+      eventName = eventName.trim().toLowerCase().replaceAll(" ", "_");
+    }
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+    return eventName;
+  }
 
+  private Bundle formatProperties(Properties properties) throws Exception {
```

Comment:
This should not throw an exception as I mentioned in https://github.com/segment-integrations/analytics-android-integration-firebase/pull/1#discussion_r137363288